packages/node/src/indexer/dynamic-ds.service.ts

```diff
@@ -0,0 +1,130 @@
+// Copyright 2020-2021 OnFinality Limited authors & contributors
+// SPDX-License-Identifier: Apache-2.0
+
+import { Injectable } from '@nestjs/common';
+import {
+  CustomDataSourceV0_2_0Impl,
+  isCustomDs,
+  isRuntimeDs,
+  manifestIsV0_2_1,
+  RuntimeDataSourceV0_2_0Impl,
+} from '@subql/common';
+import { SubqlDatasource, SubqlDatasourceKind } from '@subql/types';
+import { plainToClass } from 'class-transformer';
+import yaml from 'js-yaml';
+import { Transaction } from 'sequelize/types';
+import { SubqueryProject } from '../configure/project.model';
+import { getLogger } from '../utils/logger';
+import { DsProcessorService } from './ds-processor.service';
+import { StoreService } from './store.service';
+
+const logger = getLogger('dynamic-ds');
+
+const METADATA_KEY = 'dynamicDatasources';
+
+function replacer(key: string, value: any): any {
+  if (value instanceof Map) {
+    const obj: Record<string, unknown> = {};
+    for (const key of value.keys()) {
+      obj[key] = value.get(key);
+    }
+
+    return obj;
+  } else {
+    return value;
+  }
+}
+
+@Injectable()
+export class DynamicDsService {
+  constructor(
+    private readonly storeService: StoreService,
+    private readonly dsProcessorService: DsProcessorService,
+    private readonly project: SubqueryProject,
+  ) {}
+
+  async createDynamicDatasource(
+    templateName: string,
+    args: Record<string, unknown>,
+    currentBlock: number,
+    tx: Transaction,
+  ): Promise<void> {
+    const manifest = this.project.projectManifest;
+
+    if (!manifestIsV0_2_1(manifest)) {
+      logger.error(
+        `Unsupported project manifest version. Expected at least 0.2.1, received "${manifest.specVersion}"`,
+      );
+      process.exit(1);
+    }
+
+    const template = manifest.asV0_2_1.templates.find(
+      (t) => t.name === templateName,
+    );
+
+    if (!template) {
+      logger.error(
+        `Unable to find matching template in project for name: "${templateName}"`,
+      );
+      process.exit(1);
+    }
+
+    logger.info(`Creating new datasource from template: "${templateName}"`);
+
+    const ds: SubqlDatasource = { ...template, startBlock: currentBlock };
+    try {
+      if (isCustomDs(ds)) {
+        ds.processor.options = { ...ds.processor.options, ...args };
+        await this.dsProcessorService.validateCustomDs([ds]);
+      } else if (isRuntimeDs(ds)) {
+        // XXX add any modifications to the ds here
+
+        const runtimeDs = plainToClass(RuntimeDataSourceV0_2_0Impl, ds);
+        runtimeDs.validate();
+      }
+    } catch (e) {
+      logger.error(`Unable to create dynamic datasource.\n ${e.message}`);
+      process.exit(1);
+    }
+
+    await this.saveDynamicDatasource(ds, tx);
+  }
+
+  async getDynamicDatasources(): Promise<SubqlDatasource[]> {
+    const results = await this.storeService.getMetadata(METADATA_KEY);
```

Comment:
I would prefer injecting metaRepo into here. `storeservice` is for preparing the store object.